# Vehical Insurance

### Build a model to predict whether a customer would be interested in Vehicle Insurance 

In [1]:
import numpy as np
import pandas as pd
import matplotlib.pyplot as plt
import tensorflow as tf

In [47]:
dft=pd.read_csv('data//train.csv')
dft.head()

id  Gender  Age  Driving_License  Region_Code  Previously_Insured  \
0   1    Male   44                1         28.0                   0   
1   2    Male   76                1          3.0                   0   
2   3    Male   47                1         28.0                   0   
3   4    Male   21                1         11.0                   1   
4   5  Female   29                1         41.0                   1   

  Vehicle_Age Vehicle_Damage  Annual_Premium  Policy_Sales_Channel  Vintage  \
0   > 2 Years            Yes         40454.0                  26.0      217   
1    1-2 Year             No         33536.0                  26.0      183   
2   > 2 Years            Yes         38294.0                  26.0       27   
3    < 1 Year             No         28619.0                 152.0      203   
4    < 1 Year             No         27496.0                 152.0       39   

   Response  
0         1  
1         0  
2         1  
3         0  
4         0

 ##  Data Analyses

In [49]:
dft=dft.drop('id',axis=1)

In [50]:
dft['Gender'] = dft['Gender'].map( {'Female': 0, 'Male': 1} ).astype(int)

In [51]:
dft['Vehicle_Damage'] = dft['Vehicle_Damage'].map( {'Yes': 1, 'No': 0} ).astype(int)

#### Vehicle_Age : Need to add 3 columns. 1. Vage_mt_2= >2 Years, Vage_bt_12 =1-2 Year, Vage_lt_1=< 1 Year

In [48]:
dft.describe()

id            Age  Driving_License    Region_Code  \
count  381109.000000  381109.000000    381109.000000  381109.000000   
mean   190555.000000      38.822584         0.997869      26.388807   
std    110016.836208      15.511611         0.046110      13.229888   
min         1.000000      20.000000         0.000000       0.000000   
25%     95278.000000      25.000000         1.000000      15.000000   
50%    190555.000000      36.000000         1.000000      28.000000   
75%    285832.000000      49.000000         1.000000      35.000000   
max    381109.000000      85.000000         1.000000      52.000000   

       Previously_Insured  Annual_Premium  Policy_Sales_Channel  \
count       381109.000000   381109.000000         381109.000000   
mean             0.458210    30564.389581            112.034295   
std              0.498251    17213.155057             54.203995   
min              0.000000     2630.000000              1.000000   
25%              0.000000    24405.000000             29.000000   
50%              0.000000    31669.000000            133.000000   
75%              1.000000    39400.000000            152.000000   
max              1.000000   540165.000000            163.000000   

             Vintage       Response  
count  381109.000000  381109.000000  
mean      154.347397       0.122563  
std        83.671304       0.327936  
min        10.000000       0.000000  
25%        82.000000       0.000000  
50%       154.000000       0.000000  
75%       227.000000       0.000000  
max       299.000000       1.000000

In [52]:
dft.describe()

Gender            Age  Driving_License    Region_Code  \
count  381109.000000  381109.000000    381109.000000  381109.000000   
mean        0.540761      38.822584         0.997869      26.388807   
std         0.498336      15.511611         0.046110      13.229888   
min         0.000000      20.000000         0.000000       0.000000   
25%         0.000000      25.000000         1.000000      15.000000   
50%         1.000000      36.000000         1.000000      28.000000   
75%         1.000000      49.000000         1.000000      35.000000   
max         1.000000      85.000000         1.000000      52.000000   

       Previously_Insured  Vehicle_Damage  Annual_Premium  \
count       381109.000000   381109.000000   381109.000000   
mean             0.458210        0.504877    30564.389581   
std              0.498251        0.499977    17213.155057   
min              0.000000        0.000000     2630.000000   
25%              0.000000        0.000000    24405.000000   
50%              0.000000        1.000000    31669.000000   
75%              1.000000        1.000000    39400.000000   
max              1.000000        1.000000   540165.000000   

       Policy_Sales_Channel        Vintage       Response  
count         381109.000000  381109.000000  381109.000000  
mean             112.034295     154.347397       0.122563  
std               54.203995      83.671304       0.327936  
min                1.000000      10.000000       0.000000  
25%               29.000000      82.000000       0.000000  
50%              133.000000     154.000000       0.000000  
75%              152.000000     227.000000       0.000000  
max              163.000000     299.000000       1.000000

 ##  Scale the Data

In [4]:
cols_to_norm = ['Age','Annual_Premium','Policy_Sales_Channel','Vintage']
dft[cols_to_norm] = dft[cols_to_norm].apply(lambda x: (x - x.min()) / (x.max() - x.min()))
dft.head()

id  Gender       Age  Driving_License  Region_Code  Previously_Insured  \
0   1    Male  0.369231                1         28.0                   0   
1   2    Male  0.861538                1          3.0                   0   
2   3    Male  0.415385                1         28.0                   0   
3   4    Male  0.015385                1         11.0                   1   
4   5  Female  0.138462                1         41.0                   1   

  Vehicle_Age Vehicle_Damage  Annual_Premium  Policy_Sales_Channel   Vintage  \
0   > 2 Years            Yes        0.070366              0.154321  0.716263   
1    1-2 Year             No        0.057496              0.154321  0.598616   
2   > 2 Years            Yes        0.066347              0.154321  0.058824   
3    < 1 Year             No        0.048348              0.932099  0.667820   
4    < 1 Year             No        0.046259              0.932099  0.100346   

   Response  
0         1  
1         0  
2         1  
3         0  
4         0

KeyError: 'Vehicle_Age_lt_1_Year'

In [6]:
x_data = dft.drop('Response',axis=1)
labels = dft['Response']
from sklearn.model_selection import train_test_split
X_train, X_test, y_train, y_test = train_test_split(x_data,labels,test_size=0.10, random_state=101)

# ---------------------------------------------------------------------------------------------------
# Deep Learning Models

### Feature Columns Converstion

In [7]:
#Continuous Features¶
fc_age = tf.feature_column.numeric_column('Age')
fc_AP = tf.feature_column.numeric_column('Annual_Premium')
fc_psc = tf.feature_column.numeric_column('Policy_Sales_Channel')
fc_Vintage = tf.feature_column.numeric_column('Vintage')

### Data Input Function

In [8]:
feat_cols = [fc_age,fc_AP ,fc_psc,fc_Vintage]
input_func = tf.estimator.inputs.pandas_input_fn(x=X_train,y=y_train,batch_size=250,num_epochs=100,shuffle=True)

## 1. Linear Classifier

### Train the Model

In [9]:
model = tf.estimator.LinearClassifier(feature_columns=feat_cols,n_classes=2)
model.train(input_fn=input_func,steps=3000)

INFO:tensorflow:Using default config.
INFO:tensorflow:Using config: {'_is_chief': True, '_model_dir': 'C:\\Users\\SUGUNA~1\\AppData\\Local\\Temp\\tmptcs7ih2q', '_num_worker_replicas': 1, '_device_fn': None, '_session_creation_timeout_secs': 7200, '_protocol': None, '_evaluation_master': '', '_task_type': 'worker', '_service': None, '_experimental_max_worker_delay_secs': None, '_task_id': 0, '_train_distribute': None, '_master': '', '_keep_checkpoint_every_n_hours': 10000, '_session_config': allow_soft_placement: true
graph_options {
  rewrite_options {
    meta_optimizer_iterations: ONE
  }
}
, '_tf_random_seed': None, '_save_checkpoints_steps': None, '_num_ps_replicas': 0, '_eval_distribute': None, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x000001AB559FFC50>, '_save_summary_steps': 100, '_experimental_distribute': None, '_global_id_in_cluster': 0, '_save_checkpoints_secs': 600, '_log_step_count_steps': 100, '_keep_checkpoint_max': 5}
Instructions f

### Evaluate the Model

In [10]:
eval_input_func = tf.estimator.inputs.pandas_input_fn(
      x=X_test,
      y=y_test,
      batch_size=250,
      num_epochs=100,
      shuffle=False)
model.evaluate(eval_input_func)


INFO:tensorflow:Calling model_fn.
Instructions for updating:
Deprecated in favor of operator or tf.math.divide.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Starting evaluation at 2021-03-06T21:13:36Z
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from C:\Users\SUGUNA~1\AppData\Local\Temp\tmptcs7ih2q\model.ckpt-3000
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Finished evaluation at 2021-03-06-21:14:45
INFO:tensorflow:Saving dict for global step 3000: accuracy = 0.8788014, accuracy_baseline = 0.8788014, auc = 0.64990723, auc_precision_recall = 0.16049215, average_loss = 0.35947314, global_step = 3000, label/mean = 0.1211986, loss = 89.86475, precision = 0.0, prediction/mean = 0.12699264, recall = 0.0
INFO:tensorflow:Saving 'checkpoint_path' summary for global step 3000: C:\Users\SUGUNA~1\AppData\Local\Temp\tmptcs7ih2q\model.ckpt-3000


{'accuracy': 0.8788014,
 'accuracy_baseline': 0.8788014,
 'auc': 0.64990723,
 'auc_precision_recall': 0.16049215,
 'average_loss': 0.35947314,
 'global_step': 3000,
 'label/mean': 0.1211986,
 'loss': 89.86475,
 'precision': 0.0,
 'prediction/mean': 0.12699264,
 'recall': 0.0}

# 2. DNN Classifier

### Train the Model

In [11]:
dnn_model = tf.estimator.DNNClassifier(hidden_units=[10,10,10],feature_columns=feat_cols,n_classes=2)
dnn_model.train(input_fn=input_func,steps=1000)

INFO:tensorflow:Using default config.
INFO:tensorflow:Using config: {'_is_chief': True, '_model_dir': 'C:\\Users\\SUGUNA~1\\AppData\\Local\\Temp\\tmp1v3ca910', '_num_worker_replicas': 1, '_device_fn': None, '_session_creation_timeout_secs': 7200, '_protocol': None, '_evaluation_master': '', '_task_type': 'worker', '_service': None, '_experimental_max_worker_delay_secs': None, '_task_id': 0, '_train_distribute': None, '_master': '', '_keep_checkpoint_every_n_hours': 10000, '_session_config': allow_soft_placement: true
graph_options {
  rewrite_options {
    meta_optimizer_iterations: ONE
  }
}
, '_tf_random_seed': None, '_save_checkpoints_steps': None, '_num_ps_replicas': 0, '_eval_distribute': None, '_cluster_spec': <tensorflow.python.training.server_lib.ClusterSpec object at 0x000001AB5CAA99B0>, '_save_summary_steps': 100, '_experimental_distribute': None, '_global_id_in_cluster': 0, '_save_checkpoints_secs': 600, '_log_step_count_steps': 100, '_keep_checkpoint_max': 5}
INFO:tensorflo

### Evaluate the Model

In [12]:
eval_input_func = tf.estimator.inputs.pandas_input_fn(
      x=X_test,
      y=y_test,
      batch_size=10,
      num_epochs=1,
      shuffle=False)
dnn_model.evaluate(eval_input_func)

INFO:tensorflow:Calling model_fn.
INFO:tensorflow:Done calling model_fn.
INFO:tensorflow:Starting evaluation at 2021-03-06T21:14:53Z
INFO:tensorflow:Graph was finalized.
INFO:tensorflow:Restoring parameters from C:\Users\SUGUNA~1\AppData\Local\Temp\tmp1v3ca910\model.ckpt-1000
INFO:tensorflow:Running local_init_op.
INFO:tensorflow:Done running local_init_op.
INFO:tensorflow:Finished evaluation at 2021-03-06-21:15:03
INFO:tensorflow:Saving dict for global step 1000: accuracy = 0.8788014, accuracy_baseline = 0.8788014, auc = 0.70588064, auc_precision_recall = 0.20469445, average_loss = 0.33851022, global_step = 1000, label/mean = 0.1211986, loss = 3.3843029, precision = 0.0, prediction/mean = 0.12441013, recall = 0.0
INFO:tensorflow:Saving 'checkpoint_path' summary for global step 1000: C:\Users\SUGUNA~1\AppData\Local\Temp\tmp1v3ca910\model.ckpt-1000


{'accuracy': 0.8788014,
 'accuracy_baseline': 0.8788014,
 'auc': 0.70588064,
 'auc_precision_recall': 0.20469445,
 'average_loss': 0.33851022,
 'global_step': 1000,
 'label/mean': 0.1211986,
 'loss': 3.3843029,
 'precision': 0.0,
 'prediction/mean': 0.12441013,
 'recall': 0.0}

# -------------------------------------------------------------------------------------------------------------

# Machine Learning Models :

#### KNN, Logistic Regression, SVM, Decision Tree, Random Forest, Naive Bayes, Kernel SVM


## 1. KNN

In [13]:
# Find the best k to build the model with the best accuracy.
from sklearn.neighbors import KNeighborsClassifier
from sklearn import metrics
Ks = 50
mean_acc = np.zeros((Ks-1))
std_acc = np.zeros((Ks-1))
ConfustionMx = [];
for n in range(1,Ks):
    
    #Train Model and Predict  
    neigh = KNeighborsClassifier(n_neighbors = n).fit(X_train,y_train)
    yhat_k=neigh.predict(X_test)
    mean_acc[n-1] = metrics.accuracy_score(y_test, yhat_k)
    
    std_acc[n-1]=np.std(yhat_k==y_test)/np.sqrt(yhat_k.shape[0])

mean_acc

plt.plot(range(1,Ks),mean_acc,'g')
plt.fill_between(range(1,Ks),mean_acc - 1 * std_acc,mean_acc + 1 * std_acc, alpha=0.10)
plt.legend(('Accuracy ', '+/- 3xstd'))
plt.ylabel('Accuracy ')
plt.xlabel('Number of Nabors (K)')
plt.tight_layout()
plt.show()

print( "The best accuracy was with", mean_acc.max(), "with k=", mean_acc.argmax()+1) 

#Training model on the Training set
KNclassifier = KNeighborsClassifier(n_neighbors = mean_acc.argmax()+1, metric = 'minkowski', p = 2)
KNclassifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = KNclassifier.predict(X_test)
#print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)
print("KNN's Accuracy: ", accuracy_score(y_test, y_pred))

ValueError: could not convert string to float: '1-2 Year'

## 2. Logistic Regression model

In [ ]:
# Training the Logistic Regression model on the Training set
from sklearn.linear_model import LogisticRegression
classifier = LogisticRegression(random_state = 0)
classifier.fit(X_train, y_train)

# Predicting the Test set results
y_pred = classifier.predict(X_test)
#print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)
print("Logistic Regression's Accuracy: ", accuracy_score(y_test, y_pred))

## 3. SVM model 

In [ ]:
# Training the SVM model on the Training set
from sklearn.svm import SVC
classifier = SVC(kernel = 'linear', random_state = 0)
classifier.fit(X_train, y_train)

# Predicting a new result
#print(classifier.predict(sc.transform([[30,87000]])))

# Predicting the Test set results
y_pred = classifier.predict(X_test)
#print(np.concatenate((y_pred.reshape(len(y_pred),1), y_test.reshape(len(y_test),1)),1))

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix, accuracy_score
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)
print("SVM's Accuracy: ", accuracy_score(y_test, y_pred))

## 4. Decision Tree

In [ ]:
# Training the Decision Tree Classification model on the Training set
from sklearn.tree import DecisionTreeClassifier
classifier = DecisionTreeClassifier(criterion = 'entropy', random_state = 0)
classifier.fit(X_train, y_train)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix, accuracy_score
y_pred = classifier.predict(X_test)
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)
print("DecisionTrees's Accuracy: ", accuracy_score(y_test, y_pred))

## 5.Random Forest Classification model

In [ ]:
# Training the Random Forest Classification model on the Training set
from sklearn.ensemble import RandomForestClassifier
classifier = RandomForestClassifier(n_estimators = 10, criterion = 'entropy', random_state = 0)
classifier.fit(X_train, y_train)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix, accuracy_score
y_pred = classifier.predict(X_test)
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)
print("Random Forest's Accuracy: ", accuracy_score(y_test, y_pred))

## 6.Naive Bayes model

In [ ]:
# Training the Naive Bayes model on the Training set
from sklearn.naive_bayes import GaussianNB
classifier = GaussianNB()
classifier.fit(X_train, y_train)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix, accuracy_score
y_pred = classifier.predict(X_test)
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)
print("Naive Bayes's Accuracy: ", accuracy_score(y_test, y_pred))

## 7. Kernel SVM model

In [ ]:
# Training the Kernel SVM model on the Training set
from sklearn.svm import SVC
classifier = SVC(kernel = 'rbf', random_state = 0)
classifier.fit(X_train, y_train)

# Making the Confusion Matrix
from sklearn.metrics import confusion_matrix, accuracy_score
y_pred = classifier.predict(X_test)
cm = confusion_matrix(y_test, y_pred)
print(cm)
accuracy_score(y_test, y_pred)
print("Kernel SVM model's Accuracy: ", accuracy_score(y_test, y_pred))

# Test data -> 25 %
## Deep Learning Models :
#### DNN Classifier : 0.83
#### Linear Classifier : 0.84

## Machine Learning Models :
#### DecisionTrees's Accuracy:  0.72
#### Random Forest's Accuracy:  0.81
#### Naive Bayes's Accuracy:  0.83
#### Kernel SVM model's Accuracy:  0.84
#### Logistic Regression's Accuracy:  0.84
#### SVM's Accuracy:  0.84
#### KNN's Accuracy:  0.85 at k=7

# Test data -> 10 %
## Deep Learning Models :
#### DNN Classifier : 0.85
#### Linear Classifier : 0.84

## Machine Learning Models :
#### DecisionTrees's Accuracy:  0.746
#### Random Forest's Accuracy:  0.759
#### Logistic Regression's Accuracy:  0.84
#### Kernel SVM model's Accuracy:  0.855
#### SVM's Accuracy:  0.855
#### Naive Bayes's Accuracy:  0.855
### KNN's Accuracy:  0.89 at k=7

# ----------------------------------------------------------------------
# KNN's model provides better acurracy
# ----------------------------------------------------------------------